src/net/RTP/TextStream.cs

```diff
@@ -0,0 +1,117 @@
+
+        public Action<int, List<TextFormat>> OnTextFormatsNegotiatedByIndex { get; internal set; }
+
+        /// <summary>
+        /// Indicates whether this session is using text.
+        /// </summary>
+        public bool HasText
+        {
+            get
+            {
+                return LocalTrack != null && LocalTrack.StreamStatus != MediaStreamStatusEnum.Inactive
+                  || RemoteTrack != null && RemoteTrack.StreamStatus != MediaStreamStatusEnum.Inactive;
+            }
+        }
+
+        public void SendText(byte[] sample)
+        {
+            if (!sendingFormatFound)
+            {
+                NegotiatedFormat = GetSendingFormat();
+                sendingFormatFound = true;
+            }
+            SendTextFrame(NegotiatedFormat.ID, sample);
+        }
+
+        private void SendTextFrame(int payloadTypeID, byte[] buffer)
+        {
+            if (CheckIfCanSendRtpRaw())
+            {
+                if (rtpEventInProgress)
+                {
+                    logger.LogWarning($"An RTPEvent is in progress.");
+                    return;
+                }
+
+                try
+                {
+                    // Get the current time
+                    DateTime currentTime = DateTime.UtcNow;
+
+                    // Calculate time elapsed since the last frame in milliseconds
+                    uint elapsedMilliseconds = 0;
+
+                    if (_lastSendTime != DateTime.MinValue)
+                    {
+                        elapsedMilliseconds = (uint)(currentTime - _lastSendTime).TotalMilliseconds;
+                    }
+
+                    // Update the timestamp with elapsed time
+                    LocalTrack.Timestamp += elapsedMilliseconds;
+
+                    for (int index = 0; index * RTPSession.RTP_MAX_PAYLOAD < buffer.Length; index++)
+                    {
+                        int offset = (index == 0) ? 0 : (index * RTPSession.RTP_MAX_PAYLOAD);
+                        int payloadLength = (offset + RTPSession.RTP_MAX_PAYLOAD < buffer.Length) ? RTPSession.RTP_MAX_PAYLOAD : buffer.Length - offset;
+
+                        // Set the marker bit for the first packet after idle or session start
+                        int markerBit = _lastSendTime == DateTime.MinValue ? 1 : 0;
+
+                        byte[] payload = new byte[payloadLength];
+                        Buffer.BlockCopy(buffer, offset, payload, 0, payloadLength);
+
+                        // Send the RTP packet with the updated timestamp
+                        SendRtpRaw(payload, LocalTrack.Timestamp, markerBit, payloadTypeID, true);
+                    }
+
+                    // Update the last send time
+                    _lastSendTime = currentTime;
+                }
+                catch (SocketException sockExcp)
+                {
+                    logger.LogError("SocketException SendT140Frame. " + sockExcp.Message);
```

Comment:
As per #1306, use structured semantic logging instead of dynamic log message templates.